de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/concurrent/management/JupiterClient.java

```diff
@@ -7,34 +7,49 @@
 import de.fu_berlin.inf.dpp.concurrent.jupiter.Operation;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.TransformationException;
 import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.Jupiter;
+import de.fu_berlin.inf.dpp.concurrent.jupiter.internal.text.TimestampOperation;
+import de.fu_berlin.inf.dpp.session.AbstractActivityProducer;
 import de.fu_berlin.inf.dpp.session.ISarosSession;
-import java.util.HashMap;
+import de.fu_berlin.inf.dpp.util.NamedThreadFactory;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
 /** A JupiterClient manages Jupiter client docs for a single user with several paths */
-public class JupiterClient {
-
-  protected ISarosSession sarosSession;
-
-  public JupiterClient(ISarosSession sarosSession) {
-    this.sarosSession = sarosSession;
-  }
+public class JupiterClient extends AbstractActivityProducer {
 
+  private final ISarosSession sarosSession;
   /**
    * Jupiter instances for each local editor.
    *
    * @host and @client
    */
-  protected final HashMap<SPath, Jupiter> clientDocs = new HashMap<SPath, Jupiter>();
+  private final Map<SPath, Jupiter> clientDocs = new ConcurrentHashMap<>();
+
+  private final ScheduledThreadPoolExecutor jupiterHeartbeat =
+      new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("dpp-jupiter-heartbeat"));
+
+  public JupiterClient(ISarosSession sarosSession) {
+    this.sarosSession = sarosSession;
+    jupiterHeartbeat.scheduleWithFixedDelay(
+        () -> {
+          clientDocs.forEach(
+              (path, jupiter) -> {
+                JupiterActivity heartbeat =
+                    jupiter.generateJupiterActivity(
+                        new TimestampOperation(), sarosSession.getLocalUser(), path);
+                fireActivity(heartbeat);
+              });
+        },
+        1,
+        1,
+        TimeUnit.MINUTES);
+  }
 
   /** @host and @client */
   protected synchronized Jupiter get(SPath path) {
-
-    Jupiter clientDoc = this.clientDocs.get(path);
-    if (clientDoc == null) {
-      clientDoc = new Jupiter(true);
-      this.clientDocs.put(path, clientDoc);
-    }
-    return clientDoc;
+    return clientDocs.computeIfAbsent(path, (key) -> new Jupiter(true));
```

Comment:
This is a refactor. Should not be part of this patch.

Reply:
It is not really. The `ConcurrentHashMap` does not solve the locking problem, if I cannot atomically insert and receive the value. Which is why I swapped it to `computeIfAbsent`, which does exactly that.

Reply:
Every method in this class is synchronized. Using a ConcurrentHashMap does not gain you anything.

Reply:
Okay fair enough. I guess I can rely on synchronizing the whole object instead and maybe try to refactor this later to use a `ConcurrentHashMap` instead.